src/js_native_api_v8_internals.h

```diff
@@ -28,6 +28,45 @@
 
 namespace v8impl {
 
+class RefTracker {
+ public:
+  RefTracker(): next(nullptr), prev(nullptr) {}
+  virtual ~RefTracker() {}
+  virtual void Finalize(bool isEnvTeardown) {}
+
+  typedef RefTracker RefList;
+
+  inline void Link(RefList* list) {
+    prev = list;
+    next = list->next;
+    if (next != nullptr) {
+      next->prev = this;
+    }
+    list->next = this;
+  }
+
+  inline void Unlink() {
+    if (prev != nullptr) {
+      prev->next = next;
+    }
+    if (next != nullptr) {
+      next->prev = prev;
+    }
+    prev = nullptr;
+    next = nullptr;
+  }
+
+  static void FinalizeAll(RefList* list) {
+    while (list->next != nullptr) {
+      list->next->Finalize(true);
+    }
+  }
+
+ private:
+  RefList* next;
+  RefList* prev;
```

Comment:
also style nit: `next_` and `prev_` for non-static private members (and preferably `= nullptr` initializers here rather than in the constructor)